pkg/network/master/subnets.go

```diff
@@ -231,81 +242,131 @@ func (master *OsdnMaster) clearInitialNodeNetworkUnavailableCondition(origNode *
+	glog.V(5).Infof("Watch %s event for HostSubnet %q", watch.Deleted, hs.Name)
 
-	usedNodeIP, err := master.addNode(node.Name, string(node.UID), nodeIP, nil)
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Error creating subnet for node %s, ip %s: %v", node.Name, nodeIP, err))
+	if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; ok {
 		return
 	}
-	master.hostSubnetNodeIPs[node.UID] = usedNodeIP
-}
 
-func (master *OsdnMaster) handleDeleteNode(obj interface{}) {
-	node := obj.(*kapi.Node)
-	glog.V(5).Infof("Watch %s event for Node %q", watch.Deleted, node.Name)
-	delete(master.hostSubnetNodeIPs, node.UID)
+	master.releaseNetwork(hs.Subnet)
+}
 
-	if err := master.deleteNode(node.Name); err != nil {
-		utilruntime.HandleError(fmt.Errorf("Error deleting node %s: %v", node.Name, err))
+func (master *OsdnMaster) releaseNetwork(subnet string) {
+	_, ipnet, err := net.ParseCIDR(subnet)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error parsing subnet %q: %v", subnet, err))
 		return
 	}
-}
 
-// Watch for all hostsubnet events and if one is found with the right annotation, use the SubnetAllocator to dole a real subnet
-// This is mainly to handle F5 use case, allocate/release subnet with no real node in the cluster.
-// - Admin manually creates HostSubnet with 'AssignHostSubnetAnnotation' to allocate a subnet.
-// - Admin manually deletes HostSubnet to release the allocated subnet because there won't be
-//   node deletion event to trigger HostSubnet deletion in this case.
-func (master *OsdnMaster) watchSubnets() {
-	funcs := common.InformerFuncs(&networkapi.HostSubnet{}, master.handleAddOrUpdateSubnet, master.handleDeleteSubnet)
-	master.hostSubnetInformer.Informer().AddEventHandler(funcs)
+	for _, cn := range master.networkInfo.ClusterNetworks {
+		if cn.ClusterCIDR.Contains(ipnet.IP) {
+			sa, ok := master.subnetAllocatorMap[cn]
+			if !ok || sa == nil {
+				utilruntime.HandleError(fmt.Errorf("Subnet allocator not found for cluster network: %v", cn))
+				return
+			}
+			if err = sa.ReleaseNetwork(ipnet); err != nil {
+				utilruntime.HandleError(fmt.Errorf("Failed to release subnet %q: %v", subnet, err))
+				return
+			}
+			break
+		}
+	}
 }
 
-func (master *OsdnMaster) handleAddOrUpdateSubnet(obj, _ interface{}, eventType watch.EventType) {
-	hs := obj.(*networkapi.HostSubnet)
-	glog.V(5).Infof("Watch %s event for HostSubnet %q", eventType, hs.Name)
+// reconcileHostSubnet verifies and corrects the state of the hostsubnet.
+// Because openshift watches on events to keep hostsubnets and nodes in the correct state, missing an event
+// can cause orphaned or unusable hostsubnets to stick around.
+func (master *OsdnMaster) reconcileHostSubnet(subnet *networkapi.HostSubnet) error {
+	nodes, err := master.nodeInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error in initializing/fetching nodes: %v", err)
+	}
+	nodeUIDmap := make(map[string]string)
```

Comment:
You don't need to List all the nodes and then build nodeUIDmap. You only care about the node named subnet.Name, so just Get that one node.